File: templates/problems/14_dp_2d/04_coin_change_ii/solution.go
package coin_change_ii

// solveChange returns the number of combinations to make the amount. O(amount * len(coins)) time, O(amount) space.
func solveChange(amount int, coins []int) int {
	dp := make([]int, amount+1)
	dp[0] = 1
	for _, coin := range coins {
		for j := coin; j <= amount; j++ {
			dp[j] += dp[j-coin]
		}
	}
	return dp[amount]
}

File: templates/problems/14_dp_2d/04_coin_change_ii/python/solution.py
def solve_change(amount: int, coins: list[int]) -> int:
    """Return the number of combinations to make the amount.

    Time: O(amount * len(coins)), Space: O(amount)
    """
    dp = [0] * (amount + 1)
    dp[0] = 1
    for coin in coins:
        for j in range(coin, amount + 1):
            dp[j] += dp[j - coin]
    return dp[amount]
